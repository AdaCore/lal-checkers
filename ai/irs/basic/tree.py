"""
Provides the tree constructors of the Basic intermediate representation.
"""

from ai.utils import Bunch


def _visitable(name):
    """
    A class decorator that adds to a node class an entry point for visitors of
    this node. The given method name will be called on the visitor with the
    visited node.

    :param str name: The name of the method used to visit this type of node.

    :return: A function which can be called on a class to augment it with
        visiting capabilities.

    :rtype: type -> type
    """
    def enter_visitor(node, visitor, *args):
        """
        :param Node node: The visited node.
        :param visitors.Visitor visitor: The node visitor.
        :param *object args: Additional arguments for the visitor.
        :return: Whatever the visitor returns.
        :rtype: object
        """
        return getattr(visitor, name)(node, *args)

    def wrapper(cls):
        """
        :param type cls: The class being augmented with visiting capabilities.

        :return: The same class, with an additional method allowing visitors
            to visit its instances.

        :rtype: type
        """
        cls.visit = enter_visitor
        return cls

    return wrapper


class Node(object):
    """
    The base class for any Basic tree node.
    """
    def __init__(self, **data):
        """
        Initializes the node with the given user data.

        :param **object data: User-defined data.
        """
        self.data = Bunch(**data)

    def children(self):
        """
        :return: The children of this node.
        :rtype: iterable[Node]
        """
        raise NotImplementedError


@_visitable("visit_program")
class Program(Node):
    """
    The node which represents a "program".

    It consists of a list of statements that are to be executed one after
    the other.
    """
    def __init__(self, stmts, **data):
        """
        :param list[Stmt] stmts: The list of statements this program consists
            of.

        :param **object data: User-defined data.
        """
        Node.__init__(self, **data)
        self.stmts = stmts

    def children(self):
        for stmt in self.stmts:
            yield stmt


@_visitable("visit_ident")
class Identifier(Node):
    """
    An identifier, like "x", etc.
    """
    def __init__(self, var, **data):
        """
        :param Variable var: The variable object associated to this identifier.
        :param **object data: User-defined data.
        """
        Node.__init__(self, **data)
        self.var = var

    def children(self):
        yield self.var


@_visitable("visit_var")
class Variable(Node):
    """
    Represents a unique variable object, that can be referred to by multiple
    identifiers.
    """
    def __init__(self, name, **data):
        """
        :param string name: The name of the variable
        :param **object data: User-defined data.
        """
        Node.__init__(self, **data)
        self.name = name

    def children(self):
        return iter(())


class Stmt(Node):
    """
    The base class for all statements.
    """
    pass


@_visitable("visit_assign")
class AssignStmt(Stmt):
    """
    Represents the assign statement, i.e. [identifier] = [expr].
    """
    def __init__(self, id, expr, **data):
        """
        :param Identifier id: The identifier being assigned.
        :param Expr expr: The expression assigned to the identifier.
        :param **object data: User-defined data.
        """
        Node.__init__(self, **data)
        self.id = id
        self.expr = expr

    def children(self):
        yield self.id
        yield self.expr


@_visitable("visit_split")
class SplitStmt(Stmt):
    """
    A control-flow statement representing a nondeterministic choice of
    execution path, i.e. the two branches are visited.
    """
    def __init__(self, branches, **data):
        """
        :param list[list[Stmt]] branches: The branches of the split statement,
            where each branch is a list of statements. At least two branches
            are expected.

        :param **object data: User-defined data.
        """
        assert len(branches) >= 2

        Node.__init__(self, **data)
        self.branches = branches

    def children(self):
        for branch in self.branches:
            for stmt in branch:
                yield stmt


@_visitable("visit_loop")
class LoopStmt(Stmt):
    """
    Represents a nondeterministic loop.
    """
    def __init__(self, stmts, **data):
        """
        :param list[Stmt] stmts: The list of statements appearing in the body
            of this loop.

        :param **object data: User-defined data.
        """
        Node.__init__(self, **data)
        self.stmts = stmts

    def children(self):
        for stmt in self.stmts:
            yield stmt


@_visitable("visit_label")
class LabelStmt(Stmt):
    """
    Represents a point in the program which can be jumped to.
    """
    def __init__(self, name, **data):
        """
        :param str name: The name of the label.
        :param **object data: User-defined data.
        """
        Node.__init__(self, **data)
        self.name = name

    def children(self):
        return iter(())


@_visitable("visit_goto")
class GotoStmt(Stmt):
    """
    Represents a jump statement.
    """
    def __init__(self, label, **data):
        """
        :param LabelStmt label: The label to jump to.
        :param **object data: User-defined data.
        """
        Node.__init__(self, **data)
        self.label = label

    def children(self):
        yield self.label


@_visitable("visit_read")
class ReadStmt(Stmt):
    """
    Represents the havoc operation on a variable.
    """
    def __init__(self, id, **data):
        """
        :param Identifier id: The identifier being read.
        :param **object data: User-defined data.
        """
        Node.__init__(self, **data)
        self.id = id

    def children(self):
        yield self.id


@_visitable("visit_use")
class UseStmt(Stmt):
    """
    Represents the fact that a variable was used at this point.
    """
    def __init__(self, id, **data):
        """
        :param Identifier id: The identifier being used.
        :param **object data: User-defined data.
        """
        Node.__init__(self, **data)
        self.id = id

    def children(self):
        yield self.id


@_visitable("visit_assume")
class AssumeStmt(Stmt):
    """
    Represents the fact that an expression is assumed to be true at this point.
    """
    def __init__(self, expr, **data):
        """
        :param Expr expr: The expression being assumed.
        :param **object data: User-defined data.
        """
        Node.__init__(self, **data)
        self.expr = expr

    def children(self):
        yield self.expr


class Expr(Node):
    """
    Base class for expression nodes.
    """
    pass


@_visitable("visit_funcall")
class FunCall(Expr):
    """
    Represents a function call.
    """
    def __init__(self, fun_id, args, **data):
        """
        :param object fun_id: An object identifying the function called.
        :param list[Expr] args: The arguments passed.
        :param **object data: User-defined data.
        """
        Node.__init__(self, **data)
        self.fun_id = fun_id
        self.args = args

    def children(self):
        for arg in self.args:
            yield arg


@_visitable("visit_lit")
class Lit(Expr):
    """
    Represents a literal value.
    """
    def __init__(self, val, **data):
        """
        :param object val: Any value.
        :param **object data: User-defined data.
        """
        Node.__init__(self, **data)
        self.val = val

    def children(self):
        return iter(())
