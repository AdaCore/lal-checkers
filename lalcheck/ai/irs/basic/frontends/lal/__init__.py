"""
Provides a libadalang frontend for the Basic IR.
"""

import libadalang as lal
from lalcheck.ai.utils import profile
from lalcheck.tools.logger import log_stdout, log

import analysis
import typers
import utils
import time
from codegen import ConvertUniversalTypes, gen_ir


unknown_typer = typers.unknown_typer


class ExtractionContext(object):
    """
    The libadalang-based frontend interface. Provides method for extracting
    IR programs from Ada source files (see extract_programs), as well as
    a default typer for those programs (see default_typer).

    Note: programs extracted using different ExtractionContext are not
    compatible. Also, this extraction context must be kept alive as long
    as the programs parsed with it are intended to be used.
    """
    def __init__(self, lal_ctx=None):
        self.lal_ctx = lal.AnalysisContext() if lal_ctx is None else lal_ctx

        # Get a dummy node, needed to call static properties of libadalang.
        dummy = self.lal_ctx.get_from_buffer(
            "<dummy>", 'package Dummy is end;'
        ).root

        # Find the Character TypeDecl.
        char_type = dummy.p_standard_unit.root.find(
            lambda x: x.is_a(lal.TypeDecl) and x.f_name.text == "Character"
        )

        float_type = dummy.p_standard_unit.root.find(
            lambda x: x.is_a(lal.TypeDecl) and x.f_name.text == "Float"
        )

        self.evaluator = utils.ConstExprEvaluator(
            dummy.p_bool_type,
            dummy.p_int_type,
            float_type,
            char_type,
            dummy.p_universal_int_type,
            dummy.p_universal_real_type
        )

        self.type_models = {}
        self.fun_models = {}
        self.subpdata = {}
        self._internal_typer = self.default_typer()

    @staticmethod
    def for_project(project_file, scenario_vars={}):
        return ExtractionContext(lal.AnalysisContext(
            unit_provider=lal.UnitProvider.for_project(
                project_file,
                scenario_vars
            )
        ))

    @staticmethod
    def empty():
        return ExtractionContext()

    def extract_programs_from_file(self, ada_file):
        """
        :param str ada_file: A path to the Ada source file from which to
            extract programs.

        :return: a Basic IR Program for each subprogram body that exists in the
            given source code.

        :rtype: iterable[irt.Program]
        """
        return self._extract_from_unit(self.lal_ctx.get_from_file(ada_file))

    def extract_programs_from_provider(self, name, kind):
        return self._extract_from_unit(self.lal_ctx.get_from_provider(
            name, kind
        ))

    def extract_programs_from_unit(self, unit):
        """
        :param lal.AnalysisUnit unit: The already parsed compilation unit.
        """
        return self._extract_from_unit(unit)

    @profile()
    def use_model(self, name):
        model_unit = self.lal_ctx.get_from_provider(name, "specification")
        with log_stdout('error'):
            for diag in model_unit.diagnostics:
                print('   {}'.format(diag))
                return

        model_ofs = model_unit.root.findall(
            lambda x: (x.is_a(lal.AspectAssoc)
                       and x.f_id.text.lower() == "model_of")
        )

        type_models = {
            aspect.parent.parent.parent: aspect.f_expr.p_referenced_decl
            for aspect in model_ofs
            if aspect.parent.parent.parent.is_a(lal.TypeDecl, lal.SubtypeDecl)
        }

        fun_models = {
            aspect.parent.parent.parent: aspect.f_expr.p_referenced_decl
            for aspect in model_ofs
            if aspect.parent.parent.parent.is_a(lal.SubpDecl)
        }

        if len(type_models) + len(fun_models) != len(model_ofs):
            with log_stdout('info'):
                print("warning: detected usage of Model_Of in an unknown "
                      "context.")

        for tdecl, ref in type_models.iteritems():
            if ref is None:
                with log_stdout('info'):
                    print("warning: Model_Of used on '{}' refers to an unknown"
                          " type.".format(tdecl.f_name.text))
            else:
                self.type_models[ref] = tdecl

        for fdecl, ref in fun_models.iteritems():
            if ref is None:
                with log_stdout('info'):
                    print(
                        "warning: Model_Of used on '{}' refers to an unknown "
                        "procedure/function.".format(
                            fdecl.f_subp_spec.f_subp_name.text
                        )
                    )
            else:
                self.fun_models[ref] = fdecl

    @profile()
    def _extract_from_unit(self, unit):
        if unit.root is None:
            with log_stdout('error'):
                print('Could not parse {}:'.format(unit.filename))
                for diag in unit.diagnostics:
                    print('   {}'.format(diag))
                    return

        unit.populate_lexical_env()

        subpdata = analysis.traverse_unit(unit.root)
        self.subpdata.update(subpdata)

        progs = []
        for subp, subpuserdata in subpdata.iteritems():
            if subp.is_a(lal.BaseSubpBody):
                start_t = time.clock()
                progs.append(
                    gen_ir(self, subp, self._internal_typer, subpuserdata)
                )
                end_t = time.clock()
                log(
                    'timings',
                    " - Transformation of subprocedure {} took {}s".format(
                        subp.f_subp_spec.f_subp_name.text,
                        end_t - start_t
                    )
                )

        converter = ConvertUniversalTypes(self.evaluator, self._internal_typer)

        for prog in progs:
            prog.visit(converter)

        return progs

    def standard_typer(self):
        """
        :return: A Typer for Ada standard types of programs parsed using
            this extraction context.

        :rtype: types.Typer[lal.AdaNode]
        """
        return typers.standard_typer(self)

    def model_typer(self, inner_typer):
        return typers.model_typer(self, inner_typer)

    def default_typer(self, fallback_typer=None):
        """
        :return: The default Typer for Ada programs parsed using this
            extraction context.

        :rtype: types.Typer[lal.AdaNode]
        """
        return typers.default_typer(self, fallback_typer)
