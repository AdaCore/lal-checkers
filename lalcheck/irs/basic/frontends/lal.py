"""
Provides a libadalang frontend for the Basic IR.
"""

import libadalang as lal

from lalcheck.irs.basic import tree as irt
from lalcheck.irs.basic.visitors import ImplicitVisitor as IRImplicitVisitor
from lalcheck.irs.basic.tools import PrettyPrinter
from lalcheck.constants import ops, lits
from lalcheck import types

from funcy.calc import memoize


_lal_op_type_2_symbol = {
    (lal.OpLt, 2): irt.bin_ops[ops.Lt],
    (lal.OpLte, 2): irt.bin_ops[ops.Le],
    (lal.OpEq, 2): irt.bin_ops[ops.Eq],
    (lal.OpNeq, 2): irt.bin_ops[ops.Neq],
    (lal.OpGte, 2): irt.bin_ops[ops.Ge],
    (lal.OpGt, 2): irt.bin_ops[ops.Gt],
    (lal.OpPlus, 2): irt.bin_ops[ops.Plus],
    (lal.OpMinus, 2): irt.bin_ops[ops.Minus],
    (lal.OpDoubleDot, 2): irt.bin_ops[ops.DotDot],

    (lal.OpMinus, 1): irt.un_ops[ops.Neg],
    (lal.OpNot, 1): irt.un_ops[ops.Not],
}

_attr_2_unop = {
    'Access': irt.un_ops[ops.Address],
    'First': irt.un_ops[ops.GetFirst],
    'Last': irt.un_ops[ops.GetLast],
}


def _gen_ir(subp):
    """
    Generates Basic intermediate representation from a lal subprogram body.

    :param lal.SubpBody subp: The subprogram body from which to generate IR.

    :return: a Basic Program.

    :rtype: irt.Program
    """

    var_decls = {}

    def transform_operator(lal_op, arity):
        return _lal_op_type_2_symbol[type(lal_op), arity]

    def unimplemented(node):
        raise NotImplementedError(
            'Cannot transform "{}" ({})'.format(node.text, type(node))
        )

    def transform_expr(expr, ctx):
        if expr.is_a(lal.ParenExpr):
            return transform_expr(expr.f_expr, ctx)

        if expr.is_a(lal.BinOp):
            def lhs_ctx(lhs):
                def rhs_ctx(rhs):
                    return ctx(irt.BinExpr(
                        lhs,
                        transform_operator(expr.f_op, 2),
                        rhs,
                        type_hint=expr.p_expression_type
                    ))

                return transform_expr(expr.f_right, rhs_ctx)

            return transform_expr(expr.f_left, lhs_ctx)

        elif expr.is_a(lal.UnOp):
            return transform_expr(expr.f_expr, lambda operand: ctx(
                irt.UnExpr(
                    transform_operator(expr.f_op, 1),
                    operand,
                    type_hint=expr.p_expression_type
                )
            ))

        elif expr.is_a(lal.IfExpr):
            def cond_ctx(cond):
                def then_ctx(thn):
                    def else_ctx(els):
                        not_cond = irt.UnExpr(
                            irt.un_ops[ops.Not],
                            cond,
                            type_hint=cond.data.type_hint
                        )

                        then_stmts = [irt.AssumeStmt(cond)]
                        else_stmts = [irt.AssumeStmt(not_cond)]

                        then_stmts.extend(ctx(thn))
                        else_stmts.extend(ctx(els))

                        return [irt.SplitStmt(then_stmts, else_stmts)]

                    return transform_expr(expr.f_else_expr, else_ctx)

                return transform_expr(expr.f_then_expr, then_ctx)

            return transform_expr(expr.f_cond_expr, cond_ctx)

        elif expr.is_a(lal.Identifier):
            ref = expr.p_referenced_decl
            if ref.is_a(lal.ObjectDecl):
                return ctx(var_decls[ref, expr.text])
            elif ref.is_a(lal.EnumLiteralDecl):
                return ctx(irt.Lit(
                    expr.text,
                    type_hint=ref.parent.parent
                ))
            elif ref.is_a(lal.NumberDecl):
                return transform_expr(ref.f_expr, ctx)
            elif ref.is_a(lal.TypeDecl):
                if ref.f_type_def.is_a(lal.SignedIntTypeDef):
                    return transform_expr(ref.f_type_def.f_range.f_range, ctx)

        elif expr.is_a(lal.IntLiteral):
            return ctx(irt.Lit(
                int(expr.f_tok.text),
                type_hint=expr.p_expression_type
            ))

        elif expr.is_a(lal.NullLiteral):
            return ctx(irt.Lit(
                lits.Null,
                type_hint=expr.p_expression_type
            ))

        elif expr.is_a(lal.ExplicitDeref):
            def prefix_ctx(prefix):
                assumed_expr = irt.BinExpr(
                    prefix,
                    irt.bin_ops[ops.Neq],
                    irt.Lit(
                        lits.Null,
                        type_hint=expr.f_prefix.p_expression_type
                    ),
                    type_hint=expr.p_bool_type
                )
                assume_stmt = irt.AssumeStmt(
                    assumed_expr,
                    purpose={
                        'kind': 'deref_check',
                        'obj': prefix
                    }
                )
                return [assume_stmt] + ctx(
                    irt.UnExpr(
                        irt.un_ops[ops.Deref],
                        prefix,
                        type_hint=expr.p_expression_type
                    )
                )

            return transform_expr(expr.f_prefix, prefix_ctx)

        elif expr.is_a(lal.AttributeRef):
            return transform_expr(expr.f_prefix, lambda prefix: ctx(
                irt.UnExpr(
                    _attr_2_unop[expr.f_attribute.text],
                    prefix,
                    type_hint=expr.p_expression_type
                )
            ))

        unimplemented(expr)

    def transform_decl(decl):
        if decl.is_a(lal.TypeDecl, lal.EnumTypeDecl, lal.NumberDecl):
            return []
        elif decl.is_a(lal.ObjectDecl):
            tdecl = decl.f_type_expr.p_designated_type_decl

            for var_id in decl.f_ids:
                var_decls[decl, var_id.text] = irt.Identifier(
                    var_id.text,
                    type_hint=tdecl
                )

            if decl.f_default_expr is None:
                return [irt.ReadStmt(var_decls[decl, var_id.text])
                        for var_id in decl.f_ids]
            else:
                return transform_expr(decl.f_default_expr, lambda def_val: [
                    irt.AssignStmt(var_decls[decl, var_id.text], def_val)
                    for var_id in decl.f_ids
                ])

        unimplemented(decl)

    def transform_stmt(stmt):
        if stmt.is_a(lal.AssignStmt):
            return transform_expr(stmt.f_expr, lambda expr: [
                irt.AssignStmt(
                    var_decls[stmt.f_dest.p_referenced_decl, stmt.f_dest.text],
                    expr
                )
            ])

        elif stmt.is_a(lal.IfStmt):
            def cond_ctx(cond):
                not_cond = irt.UnExpr(
                    irt.un_ops[ops.Not],
                    cond,
                    type_hint=cond.data.type_hint
                )

                then_stmts = [irt.AssumeStmt(cond)]
                else_stmts = [irt.AssumeStmt(not_cond)]

                then_stmts.extend(transform_stmts(stmt.f_then_stmts))
                else_stmts.extend(transform_stmts(stmt.f_else_stmts))

                # todo
                # for sub in stmt.f_alternatives:
                #    traverse_branch(sub, nulls, neg_cond=stmt.f_cond_expr)

                return [irt.SplitStmt(then_stmts, else_stmts)]

            return transform_expr(stmt.f_cond_expr, cond_ctx)

        elif stmt.is_a(lal.CaseStmt):
            # todo
            return []

        elif stmt.is_a(lal.LoopStmt):
            return [irt.LoopStmt(transform_stmts(stmt.f_stmts))]

        elif stmt.is_a(lal.WhileLoopStmt):
            return transform_expr(stmt.f_spec.f_expr, lambda cond: [
                irt.LoopStmt(
                    [irt.AssumeStmt(cond)] + transform_stmts(stmt.f_stmts)
                )
            ])

        elif stmt.is_a(lal.ForLoopStmt):
            # todo
            return []

        elif stmt.is_a(lal.ExceptionHandler):
            # todo ?
            return []

        unimplemented(stmt)

    def transform_decls(decls):
        res = []
        for decl in decls:
            res.extend(transform_decl(decl))
        return res

    def transform_stmts(stmts):
        res = []
        for stmt in stmts:
            res.extend(transform_stmt(stmt))
        return res

    return irt.Program(
        transform_decls(subp.f_decls.f_decls) +
        transform_stmts(subp.f_stmts.f_stmts)
    )


class ConvertUniversalTypes(IRImplicitVisitor):
    """
    Visitor that mutates the given IR tree so as to remove references to
    universal types from in node data's type hints.
    """

    def __init__(self, unit):
        """
        :param lal.AbstractNode unit: Any lal node.
        """
        super(ConvertUniversalTypes, self).__init__()
        self.bool = unit.p_bool_type
        self.universal_int = unit.p_universal_int_type
        self.universal_real = unit.p_universal_real_type
        self.eval = ConstExprEvaluator(unit).eval

    def has_universal_type(self, expr):
        """
        :param irt.Expr expr: A Basic IR expression.

        :return: True if the expression is either of universal int type, or
            universal real type.

        :rtype: bool
        """
        return expr.data.type_hint in [self.universal_int, self.universal_real]

    def try_convert_expr(self, expr, expected_type):
        """
        :param irt.Expr expr: A Basic IR expression.

        :param lal.BaseTypeDecl expected_type: The expected type hint of the
            expression.

        :return: An equivalent expression which does not have an universal
            type.

        :rtype: irt.Expr
        """
        try:
            return irt.Lit(
                self.eval(expr),
                type_hint=expected_type
            )
        except NotConstExprError:
            expr.visit(self)
            return expr

    def visit_assign(self, assign):
        assign.expr = self.try_convert_expr(
            assign.expr,
            assign.var.data.type_hint
        )

    def visit_assume(self, assume):
        assume.expr = self.try_convert_expr(assume.expr, self.bool)

    def visit_binexpr(self, binexpr):
        expected_type = (binexpr.lhs.data.type_hint
                         if self.has_universal_type(binexpr.rhs)
                         else binexpr.rhs.data.type_hint)

        binexpr.lhs = self.try_convert_expr(binexpr.lhs, expected_type)
        binexpr.rhs = self.try_convert_expr(binexpr.rhs, expected_type)

    def visit_unexpr(self, unexpr):
        unexpr.expr.visit(self)


class NotConstExprError(ValueError):
    def __init__(self):
        super(NotConstExprError, self).__init__()


AdaTrue = 'True'
AdaFalse = 'False'


class ConstExprEvaluator(IRImplicitVisitor):
    """
    Used to evaluate expressions statically.
    See eval.
    """

    class Range(object):
        def __init__(self, first, last):
            self.first = first
            self.last = last

    BinOps = {
        ops.And: lambda x, y: ConstExprEvaluator.from_bool(
            ConstExprEvaluator.to_bool(x) and ConstExprEvaluator.to_bool(y)
        ),
        ops.Or: lambda x, y: ConstExprEvaluator.from_bool(
            ConstExprEvaluator.to_bool(x) or ConstExprEvaluator.to_bool(y)
        ),

        ops.Neq: lambda x, y: ConstExprEvaluator.from_bool(x != y),
        ops.Eq: lambda x, y: ConstExprEvaluator.from_bool(x == y),
        ops.Lt: lambda x, y: ConstExprEvaluator.from_bool(x < y),
        ops.Le: lambda x, y: ConstExprEvaluator.from_bool(x <= y),
        ops.Ge: lambda x, y: ConstExprEvaluator.from_bool(x >= y),
        ops.Gt: lambda x, y: ConstExprEvaluator.from_bool(x > y),
        ops.DotDot: lambda x, y: ConstExprEvaluator.Range(x, y),

        ops.Plus: lambda x, y: x + y,
        ops.Minus: lambda x, y: x - y
    }

    UnOps = {
        ops.Not: lambda x: ConstExprEvaluator.from_bool(
            not ConstExprEvaluator.to_bool(x)
        ),
        ops.Neg: lambda x: -x,
        ops.GetFirst: lambda x: x.first,
        ops.GetLast: lambda x: x.last
    }

    def __init__(self, unit):
        """
        :param lal.AbstractNode unit: Any lal node.
        """
        super(ConstExprEvaluator, self).__init__()
        self.bool = unit.p_bool_type
        self.int = unit.p_int_type
        self.universal_int = unit.p_universal_int_type
        self.universal_real = unit.p_universal_real_type

    @staticmethod
    def to_bool(x):
        """
        :param str x: The boolean to convert.
        :return: The representation of the corresponding boolean literal.
        :rtype: bool
        """
        return x == AdaTrue

    @staticmethod
    def from_bool(x):
        """
        :param bool x: The representation of a boolean literal to convert.
        :return: The corresponding boolean.
        :rtype: str
        """
        return AdaTrue if x else AdaFalse

    def eval(self, expr):
        """
        :param irt.Expr expr: A Basic IR expression to evaluate.
        :return: The value which this expression evalutes to.
        :rtype: int | str
        :raise NotConstExprError: if the expression is not a constant.
        :raise NotImplementedError: if implementation is incomplete.
        """
        res = self.visit(expr)
        if res is not None:
            return res
        else:
            raise NotImplementedError("Cannot evaluate `{}`".format(
                PrettyPrinter.pretty_print(expr)
            ))

    @memoize
    def visit(self, expr):
        """
        To use instead of node.visit(self). Performs memoization, so as to
        avoid evaluating expression referred to by constant symbols multiple
        times.

        :param irt.Expr expr: The IR Basic expression to evaluate

        :return: The value of this expression.

        :rtype: int | str
        """
        return expr.visit(self)

    def visit_ident(self, ident):
        raise NotConstExprError

    def visit_binexpr(self, binexpr):
        try:
            op = ConstExprEvaluator.BinOps[binexpr.bin_op.sym]
            return op(
                self.visit(binexpr.lhs),
                self.visit(binexpr.rhs)
            )
        except KeyError:
            raise NotConstExprError

    def visit_unexpr(self, unexpr):
        try:
            op = ConstExprEvaluator.UnOps[unexpr.un_op.sym]
            return op(self.visit(unexpr.expr))
        except KeyError:
            raise NotConstExprError

    def visit_lit(self, lit):
        return lit.val


@types.typer
def int_range_typer(hint):
    """
    :param lal.BaseTypeDecl hint: the lal type.
    :return: The corresponding lalcheck type.
    :rtype: types.IntRange
    """
    if hint.is_a(lal.TypeDecl):
        if hint.f_type_def.is_a(lal.SignedIntTypeDef):
            rng = hint.f_type_def.f_range.f_range
            frm = int(rng.f_left.text)
            to = int(rng.f_right.text)
            return types.IntRange(frm, to)


@types.typer
def enum_typer(hint):
    """
    :param lal.BaseTypeDecl hint: the lal type.
    :return: The corresponding lalcheck type.
    :rtype: types.Enum
    """
    if hint.is_a(lal.EnumTypeDecl):
        literals = hint.findall(lal.EnumLiteralDecl)
        return types.Enum([lit.f_enum_identifier.text for lit in literals])


def access_typer(inner_typer):
    """
    :param types.Typer[lal.BaseTypeDecl] inner_typer: A typer for elements
        being accessed.

    :return: A Typer for Ada's access types.

    :rtype: types.Typer[lal.BaseTypeDecl]
    """

    @types.typer
    def typer(hint):
        """
        :param lal.BaseTypeDecl hint: the lal type.
        :return: The corresponding lalcheck type.
        :rtype: types.Pointer
        """
        if hint.p_is_access_type:
            accessed_type = hint.f_type_def.f_subtype_indication.f_name
            tpe = inner_typer.from_hint(accessed_type.p_referenced_decl)
            if tpe:
                return types.Pointer(tpe)

    return typer


def standard_typer_of(ctx):
    """
    :param lal.AnalysisContext ctx: The lal analysis context.
    :return: A Typer for Ada's standard types.
    :rtype: types.Typer[lal.BaseTypeDecl]
    """
    node = ctx.get_from_file('standard.ads').root
    bool_type = node.p_bool_type
    int_type = node.p_int_type

    @types.typer
    def typer(hint):
        """
        :param lal.BaseTypeDecl hint: the lal type.
        :return: The corresponding lalcheck type.
        :rtype: types.Boolean | types.IntRange
        """
        if hint == bool_type:
            return types.Boolean()
        elif hint == int_type:
            return types.IntRange(-2 ** 31, 2 ** 31 - 1)

    return typer


def new_context():
    """
    Creates a new program extraction context.

    Programs extracted with the same context have compatible standard types.

    Note that the context must be kept alive as long as long as the
    programs that were extracted with this context are intended to be used.

    :return: A new libadalang analysis context.

    :rtype: lal.AnalysisContext
    """
    return lal.AnalysisContext()


def extract_programs(ctx, ada_file):
    """
    :param lal.AnalysisContext ctx: The libadalang context.

    :param str ada_file: A path to the Ada source file from which to extract
        programs.

    :return: a Basic IR Program for each subprogram body that exists in the
        given source code.

    :rtype: iterable[irt.Program]
    """
    unit = ctx.get_from_file(ada_file)

    if unit.root is None:
        print('Could not parse {}:'.format(ada_file))
        for diag in unit.diagnostics:
            print('   {}'.format(diag))
            return

    unit.populate_lexical_env()

    progs = [
        _gen_ir(subp)
        for subp in unit.root.findall((
            lal.SubpBody,
            lal.ExprFunction
        ))
    ]

    converter = ConvertUniversalTypes(unit.root)
    for prog in progs:
        prog.visit(converter)

    return progs


def default_typer(ctx):
    """
    :param lal.AnalysisContext ctx: The lal analysis context.
    :return: The default Typer for the Ada language.
    :rtype: types.Typer[lal.BaseTypeDecl]
    """

    standard_typer = standard_typer_of(ctx)

    @types.delegating_typer
    def typer():
        """
        :rtype: types.Typer[lal.BaseTypeDecl]
        """
        return (standard_typer |
                int_range_typer |
                enum_typer |
                access_typer(typer))

    return typer
