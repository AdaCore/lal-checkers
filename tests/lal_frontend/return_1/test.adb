function Ex1 return Integer is
begin
   return 42;
end Ex1;
