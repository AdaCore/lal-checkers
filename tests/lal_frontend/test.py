import lalcheck.irs.basic.frontends.lal as lal2basic
from lalcheck.irs.basic.tools import PrettyPrinter


def do(test_name, expected_output):
    ctx = lal2basic.new_context()

    progs = lal2basic.extract_programs(
        ctx,
        '{}.adb'.format(test_name)
    )

    output = PrettyPrinter.pretty_print(
        progs[0],
        PrettyPrinter.Opts(print_ids=True)
    )
    print(output)

    assert output.strip() == expected_output.strip()


do('test_constexpr_1', """
Program:
  read(x#1)
  split:
    assume(False)
    x#1 = 1
  |:
    assume(True)
""")
