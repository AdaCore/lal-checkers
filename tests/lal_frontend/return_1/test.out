Program:
  result0#1 = 42
  goto end
  end: